File: src/game/Robot.ts
import { Hand } from "./Hand"
import { History } from "./History"
import { Model } from "./Model"

export class Robot {
    private readonly model = new Model()

    constructor(private readonly nGrams: number) { }

    public nextMove(model: History): Hand {
        const nGram = model.getLast(this.nGrams)

        // Too early in the game.
        if (!nGram) { return this.fallback() }

        const prediction = this.model.get(nGram)
        if (!prediction) { return this.fallback() }

        switch (prediction) {
            case Hand.Rock: return Hand.Paper
            case Hand.Paper: return Hand.Scissors
            case Hand.Scissors: return Hand.Rock
            default: return Hand.Rock
        }
    }

    public updateModel(model: History, hand: Hand): void {
        const nGram = model.getLast(this.nGrams)
        if (nGram) { this.model.add(nGram, hand) }
    }

    private fallback(): Hand {
        return [Hand.Rock, Hand.Paper, Hand.Scissors][Math.floor(Math.random() * 3)]
    }
}

File: blazor/Robot.cs
namespace Demo
{
    public class Robot
    {
        private readonly Model model = new Model();
        private readonly int nGrams;
        private readonly Random random = new Random();

        public Robot(int nGrams)
        {
            this.nGrams = nGrams;
        }

        public Hand NextMove(History model)
        {
            var nGram = model.GetLast(this.nGrams);

            // Too early in the game.
            if (0 == nGram.Length) { return this.Fallback(); }

            var prediction = this.model.Get(nGram);
            if (prediction is null) { return this.Fallback(); }

            switch (prediction)
            {
                case Hand.Rock: return Hand.Paper;
                case Hand.Paper: return Hand.Scissors;
                case Hand.Scissors: return Hand.Rock;
                default: return Hand.Rock;
            }
        }

        public void UpdateModel(History model, Hand hand)
        {
            var nGram = model.GetLast(this.nGrams);
            if (0 < nGram.Length) { this.model.Add(nGram, hand); }
        }

        private readonly Hand[] hands = new[] { Hand.Rock, Hand.Paper, Hand.Scissors };

        private Hand Fallback()
        {
            return this.hands[(int)this.random.Next(3)];
        }
    }
}